### ISSUE TITLE
Questions About Compliance Implementation

### ISSUE THREAD

**Issue by Lhy6900** [Axellwppr/motion_tracking#6]
## Overview

Hello! While studying the compliance implementation in this project, I noticed some differences compared to the referenced `gentle-humanoid-training` project. I would like to understand the design considerations behind these differences, as I'm concerned I may have misunderstood the implementation's intent.

## Background

The project documentation mentions that the compliance method is inspired by the `gentle-humanoid-training` project. After comparing the implementations of both projects, I found some differences and would like to ask for your insights.

## Current Implementation (Motion-Tracking)

**Location**: `/active_adaptation/envs/mdp/commands/motion_tracking.py`

```python
# Compute target positions (in torso frame)
raw_target_pos_torso = quat_apply_inverse(
    self._motion.body_quat_w[:, 0, self.torso_idx_motion],
    self._motion_original.body_pos_w[:, 0] - self._motion.body_pos_w[:, 0, self.torso_idx_motion]
)

modified_target_pos_torso = quat_apply_inverse(
    self._motion.body_quat_w[:, 0, self.torso_idx_motion],
    self._motion.body_pos_w[:, 0, self.upper_force_keypoint_idx_motion] - self._motion.body_pos_w[:, 0, self.torso_idx_motion]
)

current_target_pos_torso = quat_apply_inverse(
    self.asset.data.body_link_quat_w[:, self.torso_idx_asset],
    self.asset.data.body_link_pos_w[:, self.upper_force_keypoint_idx_asset] - self.asset.data.body_link_pos_w[:, self.torso_idx_asset]
)

# Compute compliance force (two components)
upper_force_applied_torso = clamp_norm(
    (modified_target_pos_torso - raw_target_pos_torso) * self.force_kp.unsqueeze(-1),
    self.force_threshold.unsqueeze(-1)
) + clamp_norm(
    (modified_target_pos_torso - current_target_pos_torso) * 100,
    5.0
)
```

**Force composition**:
- **Part 1**: `(modified_target - raw_target) * K_p` - Force from target modification
- **Part 2**: `(modified_target - current_pos) * K_p` - Position tracking force

## Reference Implementation (Gentle-Humanoid-Training)

**Location**: `/gentle-humanoid-training/active_adaptation/envs/mdp/commands/motion_tracking.py`

```python
for _ in range(4):
    admit_point_b = self.admit.x  # current integrate pos (root frame)
    admit_point_vel_b = self.admit.v  # current integrate vel (root frame)

    # Driving force
    F_drive_b = clamp_norm(
        K_p_drive * (ref_point_b_exp - admit_point_b) +
        K_d_drive * (ref_point_vel_b_exp - admit_point_vel_b),
        max=force_limit
    )

    # External force
    F_ext_b = clamp_norm(
        self.force_kp_scaled * self.project_pos_diff(
            force_origin_b_exp - admit_point_b,
            force_dir=force_dir_b_exp,
        ),
        self.max_force * self.force_alpha
    )

    self.admit.step(F_drive_b=F_drive_b, F_ext_b=F_ext_b)
```

**Force composition**:
- **Part 1**: `K_p * (ref_point - admit_point)` - Tracking reference trajectory
- **Part 2**: `K_ext * (force_origin - admit_point)` - Responding to external forces

### Difference 1: Use of Admittance Controller

- **Gentle-Humanoid**: Uses an admittance controller (`admit`), updating the admittance point iteratively
- **Motion-Tracking**: Directly computes forces without using an admittance controller

### Difference 2: Force Computation Dependencies

This is where I'm most confused:

**Gentle-Humanoid**:
- Both force components depend on `admit_point`
- `admit_point` is dynamically updated based on the current robot state
- Therefore, both forces indirectly depend on the robot's current position `current_pos`

**Motion-Tracking**:
- Part 1 force: `(modified_target - raw_target)` - Both are given trajectory points, independent of the robot's current position
- Part 2 force: `(modified_target - current_pos)` - Depends on the robot's current position

### Regarding Learnability

In my understanding (which may be incorrect), for the reinforcement learning Actor:

**Gentle-Humanoid case**:
```
Gradient chain: action → current_pos → admit_point → F_drive, F_ext → ...
```
Both forces can backpropagate gradients, allowing the Actor to learn how to respond to these forces.

**Motion-Tracking case**:
```
Part 1: modified_target, raw_target → force_part1
        (Both are externally given, independent of current_pos)

Part 2: current_pos → force_part2
```
Part 1 force is an uncontrollable quantity for the Actor, which cannot influence it through learning.

## Request

If convenient, could you please:

1. Explain the design rationale behind this implementation approach?
2. Clarify the reasons for the differences from Gentle-Humanoid?
3. If my understanding is incorrect, please point out where I'm mistaken?

Thank you very much for your time! I'm very interested in this project and hope to better understand the design philosophy behind it.
